### PR TITLE
solana-ibc: method to update connection delay of specified connection id

### PR DESCRIPTION
**Pull request by dhruvja**
Adding a method which would enable us to change the connection delay period of a specified connection end. This method can only be called by `fee_collector` and would fail if a `connection_id` that doesnt exist is passed.

### CODE REVIEW

File: solana/solana-ibc/programs/solana-ibc/src/error.rs [hunk 1]
```diff
@@ -129,6 +129,9 @@ pub enum Error {
     /// blocks would never get finalized until more stake is added and quorum
     /// stake is less than head stake.
     MinQuorumStakeHigherThanTotalStake,
+
+    /// When a connection id which is not yet created is updated
+    InvalidConnectionId,
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/error.rs:134]: Would it make sense to use https://docs.rs/ibc/latest/ibc/core/connection/types/error/enum.ConnectionError.html#variant.ConnectionNotFound instead?

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 2]
```diff
@@ -513,6 +519,75 @@ pub mod solana_ibc {
...
+            receiver: None,
+            token_account: None,
+            token_mint: None,
+            escrow_account: None,
+            mint_authority: None,
+            token_program: None,
+            fee_collector: None,
+        };
+
+        let mut store =
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:561]: Uh? Why not storage::from_ctx!?

File: solana/solana-ibc/programs/solana-ibc/src/error.rs (hunk 1 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/error.rs:134]: i think it would.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs (hunk 2 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:561]: well then i had to rename the signer as `sender`, though it really doesnt matter what i name.

Comment (mina86) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs]: I think it’s better to rename it so we don’t have to repeat the same code.  If we *really* care about the name there might be way to make it an option in the macro, but we probably don’t care enough.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 3]
```diff
@@ -96,6 +96,12 @@ solana_program::custom_panic_default!();
 
 #[anchor_lang::program]
 pub mod solana_ibc {
+    use std::time::Duration;
+
+    use ::ibc::core::channel::context::SendPacketValidationContext;
+    use ::ibc::core::connection::types::ConnectionEnd;
+    use ::ibc::core::host::types::identifiers::ConnectionId;
+    use ::ibc::core::host::ExecutionContext;
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:104]: Those are all re-exported in `ibc` module. I’d just use `ibc::Foo` directly where those types are used.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 4]
```diff
@@ -513,6 +519,54 @@ pub mod solana_ibc {
...
+        delay_period_in_ns: u64,
+    ) -> Result<()> {
+        let storage = &mut ctx.accounts.storage;
+
+        let connection_id = ConnectionId::new(connection_id_idx);
+
+        // Panic if connection_id doenst exist
+        if storage.connections.len() >= connection_id_idx as usize {
+            return Err(error!(error::Error::ContextError(
+                crate::ibc::ContextError::ConnectionError(
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:539]: `crate::` prefix shouldn’t be needed.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 5]
```diff
@@ -513,6 +519,54 @@ pub mod solana_ibc {
...
+        ctx: Context<UpdateConnectionDelay>,
+        connection_id_idx: u64,
+        delay_period_in_ns: u64,
+    ) -> Result<()> {
+        let storage = &mut ctx.accounts.storage;
+
+        let connection_id = ConnectionId::new(connection_id_idx);
+
+        // Panic if connection_id doenst exist
+        if storage.connections.len() >= connection_id_idx as usize {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:537]: I’m not a fan of `as` casting, though I suppose there’s no obvious way
to replace it.  We could change `connection_id_idx` to `u16` and then
use `usize::from(connection_id_idx)` or do
`usize::try_from(connection_id_idx).unwrap()`.  Not sure if either of
those is actually better.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 6]
```diff
@@ -513,6 +519,54 @@ pub mod solana_ibc {
...
+                    crate::ibc::ConnectionError::ConnectionNotFound {
+                        connection_id: connection_id.clone()
+                    }
+                )
+            )));
+        }
+
+        let mut store = crate::storage::from_ctx!(ctx);
+
+        let connection_end = store.connection_end(&connection_id).unwrap();
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:549]:
```suggestion
        let connection_end = store.connection_end(&connection_id)?;
```

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 7]
```diff
@@ -513,6 +519,54 @@ pub mod solana_ibc {
...
+            connection_end.client_id().clone(),
+            connection_end.counterparty().clone(),
+            connection_end.versions().to_vec(),
+            Duration::from_nanos(delay_period_in_ns),
+        )
+        .unwrap();
+
+        store
+            .store_connection(
+                &::ibc::core::host::types::path::ConnectionPath(connection_id),
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:562]:
```suggestion
                &ibc::path::ConnectionPath(connection_id),
```
should work

**Review by mina86 — APPROVED** [review]

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs (hunk 5 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:534]: if we use u16, we would need to convert it again to get `connection_id`

Comment (mina86) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:537]: Yes, but you can do it with just `_.into()` or `u64::from(_)`.

Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:537]: but what is the benefit of getting rid of `as` casting. does it affect the performance?

Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs]: oh i see. so it makes sense to send it as u16 and use `from` and `into` for conversions then.

**Pull request merged**